File: src/main/java/no/nav/sosialhjelp/soknad/business/sendtsoknad/SendtSoknadRepositoryJdbc.java
package no.nav.sosialhjelp.soknad.business.sendtsoknad;

import no.nav.sosialhjelp.soknad.domain.SendtSoknad;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcDaoSupport;
import org.springframework.stereotype.Component;
import org.springframework.transaction.support.TransactionTemplate;

import javax.inject.Inject;
import javax.inject.Named;
import javax.sql.DataSource;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.ZoneId;
import java.util.Optional;

import static java.time.LocalDateTime.now;
import static java.util.Date.from;
import static no.nav.sosialhjelp.soknad.business.db.SQLUtils.selectNextSequenceValue;

@Named("SendtSoknadRepository")
@Component
public class SendtSoknadRepositoryJdbc extends NamedParameterJdbcDaoSupport implements SendtSoknadRepository {

    @Inject
    private TransactionTemplate transactionTemplate;

    @Inject
    public void setDS(DataSource ds) {
        super.setDataSource(ds);
    }

    @Override
    public Long opprettSendtSoknad(SendtSoknad sendtSoknad, String eier) {
        sjekkOmBrukerEierSendtSoknad(sendtSoknad, eier);
        Long sendtSoknadId = getJdbcTemplate().queryForObject(selectNextSequenceValue("SENDT_SOKNAD_ID_SEQ"), Long.class);
        getJdbcTemplate()
                .update("insert into SENDT_SOKNAD (sendt_soknad_id, behandlingsid, tilknyttetbehandlingsid, eier, fiksforsendelseid, orgnr, navenhetsnavn, brukeropprettetdato, brukerferdigdato, sendtdato)" +
                                " values (?,?,?,?,?,?,?,?,?,?)",
                        sendtSoknadId,
                        sendtSoknad.getBehandlingsId(),
                        sendtSoknad.getTilknyttetBehandlingsId(),
                        sendtSoknad.getEier(),
                        sendtSoknad.getFiksforsendelseId(),
                        sendtSoknad.getOrgnummer(),
                        sendtSoknad.getNavEnhetsnavn(),
                        from(sendtSoknad.getBrukerOpprettetDato().atZone(ZoneId.systemDefault()).toInstant()),
                        from(sendtSoknad.getBrukerFerdigDato().atZone(ZoneId.systemDefault()).toInstant()),
                        sendtSoknad.getSendtDato() != null ? from(sendtSoknad.getSendtDato().atZone(ZoneId.systemDefault()).toInstant())
                         : null);
        return sendtSoknadId;
    }

    @Override
    public Optional<SendtSoknad> hentSendtSoknad(String behandlingsId, String eier) {
        return getJdbcTemplate().query("select * from SENDT_SOKNAD where EIER = ? and BEHANDLINGSID = ?",
                new SendtSoknadRowMapper(), eier, behandlingsId).stream().findFirst();
    }

    @Override
    public void oppdaterSendtSoknadVedSendingTilFiks(String fiksforsendelseId, String behandlingsId, String eier) {
        getJdbcTemplate()
                .update("update SENDT_SOKNAD set FIKSFORSENDELSEID = ?, SENDTDATO = ? where BEHANDLINGSID = ? and EIER = ?",
                        fiksforsendelseId,
                        from(now().atZone(ZoneId.systemDefault()).toInstant()),
                        behandlingsId,
                        eier);
    }

    private void sjekkOmBrukerEierSendtSoknad(SendtSoknad sendtSoknad, String eier) {
        if (eier == null || !eier.equalsIgnoreCase(sendtSoknad.getEier())) {
            throw new RuntimeException("Eier stemmer ikke med søknadens eier");
        }
    }

    public class SendtSoknadRowMapper implements RowMapper<SendtSoknad> {

        public SendtSoknad mapRow(ResultSet rs, int rowNum) throws SQLException {
            return new SendtSoknad()
                    .withSendtSoknadId(rs.getLong("sendt_soknad_id"))
                    .withBehandlingsId(rs.getString("behandlingsid"))
                    .withTilknyttetBehandlingsId(rs.getString("tilknyttetbehandlingsid"))
                    .withEier(rs.getString("eier"))
                    .withFiksforsendelseId(rs.getString("fiksforsendelseid"))
                    .withOrgnummer(rs.getString("orgnr"))
                    .withNavEnhetsnavn(rs.getString("navenhetsnavn"))
                    .withBrukerOpprettetDato(rs.getTimestamp("brukeropprettetdato") != null ?
                            rs.getTimestamp("brukeropprettetdato").toLocalDateTime() : null)
                    .withBrukerFerdigDato(rs.getTimestamp("brukerferdigdato") != null ?
                            rs.getTimestamp("brukerferdigdato").toLocalDateTime() : null)
                    .withSendtDato(rs.getTimestamp("sendtdato") != null ?
                            rs.getTimestamp("sendtdato").toLocalDateTime() : null);
        }
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/db/repositories/sendtsoknad/SendtSoknadRepositoryJdbc.kt
package no.nav.sosialhjelp.soknad.db.repositories.sendtsoknad

import no.nav.sosialhjelp.soknad.db.SQLUtils
import no.nav.sosialhjelp.soknad.db.repositories.sendtsoknad.SendtSoknadRowMapper.sendtSoknadRowMapper
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Repository
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.Date

@Suppress("RECEIVER_NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
@Repository
open class SendtSoknadRepositoryJdbc(
    private val jdbcTemplate: JdbcTemplate,
) : SendtSoknadRepository {

    override fun opprettSendtSoknad(sendtSoknad: SendtSoknad, eier: String?): Long? {
        sjekkOmBrukerEierSendtSoknad(sendtSoknad, eier)
        val sendtSoknadId = jdbcTemplate.queryForObject(
            SQLUtils.selectNextSequenceValue("SENDT_SOKNAD_ID_SEQ"),
            Long::class.java
        )
        jdbcTemplate.update(
            "insert into SENDT_SOKNAD (sendt_soknad_id, behandlingsid, tilknyttetbehandlingsid, eier, fiksforsendelseid, orgnr, navenhetsnavn, brukeropprettetdato, brukerferdigdato, sendtdato) values (?,?,?,?,?,?,?,?,?,?)",
            sendtSoknadId,
            sendtSoknad.behandlingsId,
            sendtSoknad.tilknyttetBehandlingsId,
            sendtSoknad.eier,
            sendtSoknad.fiksforsendelseId,
            sendtSoknad.orgnummer,
            sendtSoknad.navEnhetsnavn,
            Date.from(sendtSoknad.brukerOpprettetDato.atZone(ZoneId.systemDefault()).toInstant()),
            Date.from(sendtSoknad.brukerFerdigDato.atZone(ZoneId.systemDefault()).toInstant()),
            sendtSoknad.sendtDato?.let { Date.from(it.atZone(ZoneId.systemDefault()).toInstant()) }
        )
        return sendtSoknadId
    }

    override fun hentSendtSoknad(behandlingsId: String, eier: String?): SendtSoknad? {
        return jdbcTemplate.query(
            "select * from SENDT_SOKNAD where EIER = ? and BEHANDLINGSID = ?",
            sendtSoknadRowMapper,
            eier,
            behandlingsId
        ).firstOrNull()
    }

    override fun oppdaterSendtSoknadVedSendingTilFiks(
        fiksforsendelseId: String?,
        behandlingsId: String?,
        eier: String?,
    ) {
        jdbcTemplate.update(
            "update SENDT_SOKNAD set FIKSFORSENDELSEID = ?, SENDTDATO = ? where BEHANDLINGSID = ? and EIER = ?",
            fiksforsendelseId,
            Date.from(LocalDateTime.now().atZone(ZoneId.systemDefault()).toInstant()),
            behandlingsId,
            eier
        )
    }

    private fun sjekkOmBrukerEierSendtSoknad(sendtSoknad: SendtSoknad, eier: String?) {
        if (eier == null || !eier.equals(sendtSoknad.eier, ignoreCase = true)) {
            throw RuntimeException("Eier stemmer ikke med søknadens eier")
        }
    }
}
